### PR TITLE
EVA-681 GA4GH web services using Spring - Testing

### PR DESCRIPTION
**Pull request by tomdcsmith**
Use of SpringBootTests and Mockito to test the GA4GH webservices.

Requires latest changes to hotfix/0.4 branch of EBIvariation/biodata, including merging of the PR: https://github.com/EBIvariation/biodata/pull/6.

### CODE REVIEW

File: eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconWSServerTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                                   chromosome,
+                                   start,
+                                   allele,
+                                   String.join(",", datasetIds));
+        ResponseEntity<GA4GHBeaconResponse> response = restTemplate.getForEntity(
+                url, GA4GHBeaconResponse.class);
+
+        assertEquals(HttpStatus.OK, response.getStatusCode());
+
+        GA4GHBeaconResponse ga4GHBeaconResponse = response.getBody();
```
Comment (cyenyxe) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconWSServerTest.java]: `ga4ghBeaconResponse` as variable for more consistent capitalization.

File: eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconWSServerTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public void testAltAlleleExistsBeacon() throws Exception {
+        assertTrue(testBeaconHelper("1", 1000, "alternate", new ArrayList<>()));
+    }
+
+    @Test
+    public void testAltAlleleDoesntExistsBeacon() throws Exception {
+        assertFalse(testBeaconHelper("2", 2000, "alternateOther", new ArrayList<>()));
+    }
+
+    private Boolean testBeaconHelper(String chromosome, int start, String allele, List<String> datasetIds) {
```
Comment (cyenyxe) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconWSServerTest.java]: The return type can be the primitive `boolean`.

File: eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantCallSetWSServerTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        samplesPosition.put("sample1", 123);
+        VARIANT_SOURCE_ENTITY.setSamplesPosition(samplesPosition);
+
+        List<VariantSourceEntity> variantSourceEntities = Collections.singletonList(VARIANT_SOURCE_ENTITY);
+
+        given(variantSourceEntityRepository.findByFileIdIn(eq(Collections.singletonList("fileId")), any()))
+                .willReturn(variantSourceEntities);
+
+        given(variantSourceEntityRepository.countByFileIdIn(eq(Collections.singletonList("fileId"))))
+                .willReturn(1L);
```
Comment (cyenyxe) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantCallSetWSServerTest.java]: Replacing 1L with `variantSourceEntities.size()` will ensure that even if the collection size changes, the mocked data will make sense.

File: eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantSetWSServerTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        metadata.put("header", "myHeader");
+        VARIANT_SOURCE_ENTITY.setMetadata(metadata);
+
+        List<VariantSourceEntity> variantSourceEntities = Collections.singletonList(VARIANT_SOURCE_ENTITY);
+
+        given(variantSourceEntityRepository.findByStudyIdIn(eq(Collections.singletonList("studyId")), any()))
+                .willReturn(variantSourceEntities);
+
+        given(variantSourceEntityRepository.countByStudyIdIn(eq(Collections.singletonList("studyId"))))
+                .willReturn(1L);
```
Comment (cyenyxe) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantSetWSServerTest.java]: Replacing 1L with `variantSourceEntities.size()` will ensure that even if the collection size changes, the mocked data will make sense.

File: eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantWSServerTest.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        Region region = new Region("1", 500, 2000);
+
+        given(variantEntityRepository.findByRegionsAndComplexFilters(eq(Collections.singletonList(region)),
+                                                                     any(),
+                                                                     any(),
+                                                                     any()))
+                .willReturn(variantEntities);
+        given(variantEntityRepository.countByRegionsAndComplexFilters(eq(Collections.singletonList(region)),
+                                                                     any()))
+                .willReturn(new Long(variantEntities.size()));
```
Comment (cyenyxe) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantWSServerTest.java]: It is preferred to use `Long.valueOf()` over `new Long()` for casts. In this case it won't affect memory usage much, but given that it is the most common idiom, I would still recommend replacing it.

File: eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconResponse.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    private String allele;
+
+    private String datasetIds;
+
+    private boolean exists;
+
+    private String errorMessage;
+
+    public GA4GHBeaconResponse() {
```
Comment (jorizci) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconResponse.java]: Empty constructor is required, but if I recall correctly class protected is enough.

File: eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconResponse.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    public GA4GHBeaconResponse(String chromosome, Integer start, String allele, String datasetIds, String errorMessage) {
+        this.chromosome = chromosome;
+        this.start = start;
+        this.allele = allele;
+        this.datasetIds = datasetIds;
+        this.errorMessage = errorMessage;
+    }
+
+    public String getChromosome() {
```
Comment (jorizci) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconResponse.java]: Boilerplate. Tested in file tracking. The only precaution one has to do is avoiding bidirectional class relationships. This is fixed by using a specific annotation that jackson needs to identify and handle it himself.

Comment (cyenyxe) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconResponse.java]: I don't recall any bidirectional relations in these classes.

File: eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantCallSetWSServerTest.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class GA4GHVariantCallSetWSServerTest {
+
+    private static VariantSourceEntity VARIANT_SOURCE_ENTITY =
```
Comment (jmmut) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantCallSetWSServerTest.java]: Not really my style (having an upper case class field that is modified), but I'm ok with this, given that it is only updated in the setup. Applies to other tests classes.

Comment (cyenyxe) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantCallSetWSServerTest.java]: In fact, this object doesn't need to be an attribute of the class as it is only used inside the setUp() method.

Comment (tomdcsmith) [inline eva-server/src/test/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHVariantCallSetWSServerTest.java]: I've moved this and those in the other tests to local variables now.

**Pull request merged**